chiselFrontend/src/main/scala/chisel3/internal/firrtl/IR.scala

```diff
@@ -63,49 +63,73 @@ case class Node(id: HasId) extends Arg {
   }
 }
 
-abstract class LitArg(val num: BigInt, widthArg: Width) extends Arg {
+sealed trait LitArg extends Arg {
+  def display: String = name
+  val widthArg: Width
   private[chisel3] def forcedWidth = widthArg.known
   private[chisel3] def width: Width = if (forcedWidth) widthArg else Width(minWidth)
   override def fullName(ctx: Component): String = name
+
+  protected[internal] def minWidth: Int
+  // if (forcedWidth) {
+  //   require(widthArg.get >= minWidth,
+  //     s"The literal value ${display} was elaborated with a specified width of ${widthArg.get} bits, but at least ${minWidth} bits are required.")
+  // }
+}
+
+sealed trait BitsLitArg extends LitArg {
+  val num: BigInt
+  override def display = num.toString
   // Ensure the node representing this LitArg has a ref to it and a literal binding.
   def bindLitArg[T <: Element](elem: T): T = {
     elem.bind(ElementLitBinding(this))
     elem.setRef(this)
     elem
   }
+}
 
-  protected def minWidth: Int
-  if (forcedWidth) {
-    require(widthArg.get >= minWidth,
-      s"The literal value ${num} was elaborated with a specified width of ${widthArg.get} bits, but at least ${minWidth} bits are required.")
+class AggregateLit(val litMap: Seq[(String, Data, LitArg)]) extends LitArg {
```

Comment:
Can you unify the redundant information (you have both the name and a Data, can you just have one) instead of overencoding data and having potential for consistency issues?

Reply:
I have a `Seq` because I would like the order to match the bundle- I think this is important for things like `asUInt` to work as expected, but perhaps that is incorrect. I think `asUInt` will take a literal like `{ a : UInt("h3"), b : UInt("h4") }` and use a asUInt `primop` rather than emitting the `UInt` directly, which would require the order to be correct.

Re: the redundancy, I agree it's not great. I'm not really sure what the right thing to do is, though. I need names to correctly emit firrtl, but the internal chisel stuff wants the `Data`. I think it's easier to go from names to data via the bundle API, but I wanted to get your thoughts on that.

Reply:
I think I also thought about names vs. Data ref, and my implementation uses the Data ref. I think when you're emitting, you should reference the top-level Aggregate and get names from there.

I don't think having the Seq for ordering is a good idea, the ordering information then becomes stored in two places (in the Seq ordering and in the Bundle elements ordering) and non-DRY. I think it makes sense to use the Bundle elements as the authoritative data source, with this just providing additional data.

I agree with pawning as much work to FIRRTL as possible - doing things like const prop in the frontend makes it more likely to have subtle differences in behavior. And you have to keep your implementation in sync in two places.

Reply:
My current approach is that when aggregates are bound as literals, the children get bound as literals too. I think this is a lot more straightforward than references to bundle elements look at `topBindingOpt` and then searching through a map. `Data` equality checking and `hashCode`s don't really work, right?